client/service/src/chain_ops.rs

```diff
@@ -116,55 +303,36 @@ impl<
 			let client = &self.client;
 			let queue = &mut self.import_queue;
 
-			// Start by reading the number of blocks if not done so already.
-			let count = match count {
-				Some(c) => c,
-				None => {
-					let c: u64 = match Decode::decode(&mut io_reader_input) {
-						Ok(c) => c,
-						Err(err) => {
-							let err = format!("Error reading file: {}", err);
-							return std::task::Poll::Ready(Err(From::from(err)));
-						},
-					};
-					info!("📦 Importing {} blocks", c);
-					count = Some(c);
-					c
-				}
-			};
-
 			// Read blocks from the input.
-			if read_block_count < count {
-				match SignedBlock::<Self::Block>::decode(&mut io_reader_input) {
-					Ok(signed) => {
-						let (header, extrinsics) = signed.block.deconstruct();
-						let hash = header.hash();
-						// import queue handles verification and importing it into the client
-						queue.import_blocks(BlockOrigin::File, vec![
-							IncomingBlock::<Self::Block> {
-								hash,
-								header: Some(header),
-								body: Some(extrinsics),
-								justification: signed.justification,
-								origin: None,
-								allow_missing_state: false,
-								import_existing: force,
-							}
-						]);
+			match block_iter.next() {
+				None => {
+					let read_block_count = block_iter.read_block_count();
+					let num_expected_blocks = block_iter.num_expected_blocks();
+
+					if importing_is_done(num_expected_blocks, read_block_count, link.imported_blocks) {
+						info!(
+							"🎉 Imported {} blocks. Best: #{}",
+							read_block_count, client.chain_info().best_number
+						);
+						return std::task::Poll::Ready(Ok(()))
 					}
-					Err(e) => {
-						warn!("Error reading block data at {}: {}", read_block_count, e);
-						return std::task::Poll::Ready(Ok(()));
+				},
+				Some(block_result) => {
+					let read_block_count = block_iter.read_block_count();
+					// Make sure we are not running more than MAX_PENDING_BLOCKS ahead of the queue.
+					// `read_block_count` should always be >= to `link.imported_blocks` so this is safe from underflow.
+					if read_block_count - link.imported_blocks < MAX_PENDING_BLOCKS {
+						match block_result {
+							Ok(signed_block) => import_block_to_queue(signed_block, queue, force),
+							Err(e) => {
+								let read_block_count = block_iter.read_block_count();
+								warn!("Error reading block data at {}: {}", read_block_count, e);
```

Comment:
We should return the error here.

Reply:
And not print it as warning (just below we return `Ok(())`)

Reply:
Ok. This was in the code before so I left it as is :)

Reply:
Implemented by https://github.com/paritytech/substrate/pull/5816/commits/9865fae4014c99b0d50bb809f86c5bc2aea1b48a . I've removed the warn! and have not replaced it since we are now returning the error message.